### ISSUE TITLE
Verification may be falsely rejected when root cert has authorityCertIssuer field

### ISSUE THREAD

**Issue by mxsasha** [pyca/cryptography#11461]
Cryptography's x509 verification appears to reject verification on a certificate when the authorityCertIssuer field is present in the authority key identifier extension of the root certificate. This may be correct per CABF baseline, but is suspiciously different from all other implementations I tested.

I have [written a small script that isolates the issue](https://gist.github.com/mxsasha/9ca38da42c08f7ec9d1a120f6c82f9d8). This fails on `cryptography.hazmat.bindings._rust.x509.VerificationError: validation failed: CandidatesExhausted(Other("authorityKeyIdentifier must not contain authorityCertIssuer"))`. The leaf and intermediate do not contain authorityCertIssuer - the root does. 

The error message [originates in the code here](https://github.com/pyca/cryptography/blob/7fda121e69f40b16cc8bf46f9f7ea8cf217e88cb/src/rust/cryptography-x509-verification/src/policy/extension.rs#L418). This code seems to have an exception that allows the authority key identifier extension to be absent, but if it is found, the presence of authorityCertIssuer will lead to an error.

This is a widely trusted root certificate, and no browsers and other testing tools don't seem to have an issue with it. I do see the [CABF baseline](https://cabforum.org/working-groups/server/baseline-requirements/requirements/#71213-root-ca-authority-key-identifier) does not allow authorityCertIssuer. But this was [introduced in 2020](https://github.com/cabforum/servercert/pull/195), the root certificate is from 2006. Perhaps there are other details I don't know - I'm not too familiar with these policies and their implementations. It definitely seems strange and undesirable that cryptography rejects a certificate where other implementations find it valid.

It would be great if someone with more experience in this area could look at this and determine if cryptography is too strict in this case, and what the correct behavior would be.

For completeness, my versions, with cryptography installed in a venv through pip install:
```
cffi==1.15.1
cryptography==43.0.0
pip-tools==6.14.0
setuptools==73.0.0

Python 3.9.6 (default, Nov 10 2023, 13:38:27) 
[Clang 15.0.0 (clang-1500.1.0.2.5)] on darwin
```

**Comment by alex** [comment]
cc: @woodruffw

**Comment by mxsasha** [comment]
Doing a quick scan on 147 root certificates from a local bundle, this seems rare, but there are a few others (trimmed):

```
Certificate:
    Data:
        Serial Number: 1478 (0x5c6)
        Signature Algorithm: sha1WithRSAEncryption
        Issuer: C=BM, O=QuoVadis Limited, CN=QuoVadis Root CA 3
        Validity
            Not Before: Nov 24 19:11:23 2006 GMT
            Not After : Nov 24 19:06:44 2031 GMT
        Subject: C=BM, O=QuoVadis Limited, CN=QuoVadis Root CA 3
        X509v3 extensions:
            X509v3 Basic Constraints: critical
                CA:TRUE
            X509v3 Key Usage:
                Certificate Sign, CRL Sign
            X509v3 Subject Key Identifier:
                F2:C0:13:E0:82:43:3E:FB:EE:2F:67:32:96:35:5C:DB:B8:CB:02:D0
            X509v3 Authority Key Identifier:
                keyid:F2:C0:13:E0:82:43:3E:FB:EE:2F:67:32:96:35:5C:DB:B8:CB:02:D0
                DirName:/C=BM/O=QuoVadis Limited/CN=QuoVadis Root CA 3
                serial:05:C6

Certificate:
    Data:
        Serial Number: 0 (0x0)
        Signature Algorithm: sha1WithRSAEncryption
        Issuer: C=US, O=The Go Daddy Group, Inc., OU=Go Daddy Class 2 Certification Authority
        Validity
            Not Before: Jun 29 17:06:20 2004 GMT
            Not After : Jun 29 17:06:20 2034 GMT
        Subject: C=US, O=The Go Daddy Group, Inc., OU=Go Daddy Class 2 Certification Authority
        X509v3 extensions:
            X509v3 Subject Key Identifier:
                D2:C4:B0:D2:91:D4:4C:11:71:B3:61:CB:3D:A1:FE:DD:A8:6A:D4:E3
            X509v3 Authority Key Identifier:
                keyid:D2:C4:B0:D2:91:D4:4C:11:71:B3:61:CB:3D:A1:FE:DD:A8:6A:D4:E3
                DirName:/C=US/O=The Go Daddy Group, Inc./OU=Go Daddy Class 2 Certification Authority
                serial:00
            X509v3 Basic Constraints:
                CA:TRUE

Certificate:
    Data:
        Serial Number: 0 (0x0)
        Signature Algorithm: sha1WithRSAEncryption
        Issuer: C=US, O=Starfield Technologies, Inc., OU=Starfield Class 2 Certification Authority
        Validity
            Not Before: Jun 29 17:39:16 2004 GMT
            Not After : Jun 29 17:39:16 2034 GMT
        Subject: C=US, O=Starfield Technologies, Inc., OU=Starfield Class 2 Certification Authority
        X509v3 extensions:
            X509v3 Subject Key Identifier:
                BF:5F:B7:D1:CE:DD:1F:86:F4:5B:55:AC:DC:D7:10:C2:0E:A9:88:E7
            X509v3 Authority Key Identifier:
                keyid:BF:5F:B7:D1:CE:DD:1F:86:F4:5B:55:AC:DC:D7:10:C2:0E:A9:88:E7
                DirName:/C=US/O=Starfield Technologies, Inc./OU=Starfield Class 2 Certification Authority
                serial:00
            X509v3 Basic Constraints:
                CA:TRUE
    Signature Algorithm: sha1WithRSAEncryption


Certificate:
    Data:
        Serial Number:
            fe:dc:e3:01:0f:c9:48:ff
        Signature Algorithm: sha1WithRSAEncryption
        Issuer: C=FR, O=Dhimyotis, CN=Certigna
        Validity
            Not Before: Jun 29 15:13:05 2007 GMT
            Not After : Jun 29 15:13:05 2027 GMT
        Subject: C=FR, O=Dhimyotis, CN=Certigna
        X509v3 extensions:
            X509v3 Basic Constraints: critical
                CA:TRUE
            X509v3 Subject Key Identifier:
                1A:ED:FE:41:39:90:B4:24:59:BE:01:F2:52:D5:45:F6:5A:39:DC:11
            X509v3 Authority Key Identifier:
                keyid:1A:ED:FE:41:39:90:B4:24:59:BE:01:F2:52:D5:45:F6:5A:39:DC:11
                DirName:/C=FR/O=Dhimyotis/CN=Certigna
                serial:FE:DC:E3:01:0F:C9:48:FF
            X509v3 Key Usage: critical
                Certificate Sign, CRL Sign
            Netscape Cert Type:
                SSL CA, S/MIME CA, Object Signing CA
```

**Comment by woodruffw** [comment]
Thanks for the ping @alex, and for the report @mxsasha!

For posterity, the CABF section that forbids ACI in CA AKIs is `7.1.2.1.3 Root CA Authority Key Identifier`:


Field | Description
-- | --
keyIdentifier | MUST be present. MUST be identical to the subjectKeyIdentifier field.
authorityCertIssuer | MUST NOT be present
authorityCertSerialNumber | MUST NOT be present


Meanwhile, RFC 5280 4.2.1.1 is much less prescriptive:

> This profile RECOMMENDS support for the key identifier method by all
> certificate users.

x509-limbo also agrees that implementations other than PyCA's are overwhelmingly permissive of ACIs in root CA AKIs: <https://x509-limbo.com/testcases/webpki/#webpkiakiroot-with-aki-authoritycertissuer>

TL;DR: I agree that this should be relaxed, similar to what we've done with other pedantic checks 🙂. I'll look into a patch today.

(x509-limbo also has cases for root CAs with ACSNs, which PyCA is similarly strict about while other implementations permit it. So I'll relax that check at the same time: <https://x509-limbo.com/testcases/webpki/#webpkiakiroot-with-aki-authoritycertserialnumber>)

**Comment by woodruffw** [comment]
#11462 should address this!
